src/demand/water_demand.jl

```diff
@@ -0,0 +1,934 @@
+        if model.parameters.irrigation_areas[i]
+            paddy_runoff = max(runoff[i] - model.parameters.h_max[i], 0.0)
+            model.variables.h[i] = runoff[i] - paddy_runoff
+            runoff[i] = paddy_runoff
+        end
+    end
+    return nothing
+end
+update_runoff!(model::NoIrrigationPaddy, runoff) = nothing
+
+"""
+    update_demand_gross!(model::Paddy)
+
+Update gross water demand `demand_gross` of the paddy irrigation model for a single
+timestep.
+
+The gross water demand is based on irrigation that is applied when the `irrigation_trigger`
+is `true` (`on`) and when the paddy water depth `h` reaches below the minimum water depth
+`h_min`. Irrigation is the amount required to reach the optimal paddy water depth `h_opt`,
+taking into account limited irrigation efficiency and limited by a maximum irrigation rate.
+"""
+function update_demand_gross!(model::Paddy)
+    (;
+        irrigation_areas,
+        irrigation_trigger,
+        irrigation_efficiency,
+        maximum_irrigation_rate,
+        h_opt,
+        h_min,
+    ) = model.parameters
+    (; h, demand_gross) = model.variables
+    for i in eachindex(irrigation_areas)
+        if irrigation_areas[i] && irrigation_trigger[i]
+            # check if maximum irrigation rate has been applied at the previous time step.
+            max_irri_rate_applied = demand_gross[i] == maximum_irrigation_rate[i]
+            # start irrigation
+            irr_depth_paddy = if h[i] < h_min[i]
+                h_opt[i] - h[i]
+            elseif h[i] < h_opt[i] && max_irri_rate_applied # continue irrigation
+                h_opt[i] - h[i]
+            else
+                0.0
+            end
+            irri_dem_gross = irr_depth_paddy / irrigation_efficiency[i]
+            # limit irrigation demand to the maximum irrigation rate
+            irri_dem_gross = min(irri_dem_gross, maximum_irrigation_rate[i])
+        else
+            irri_dem_gross = 0.0
+        end
+        demand_gross[i] = irri_dem_gross
+    end
+    return nothing
+end
+
+update_demand_gross!(paddy::NoIrrigationPaddy) = nothing
+
+"Struct to store water demand model variables"
+@get_units @grid_loc @with_kw struct DemandVariables{T}
+    irri_demand_gross::Vector{T}        # irrigation gross demand [mm Δt⁻¹]
+    nonirri_demand_gross::Vector{T}     # non-irrigation gross demand [mm Δt⁻¹]
+    total_gross_demand::Vector{T}       # total gross demand [mm Δt⁻¹]
+    surfacewater_demand::Vector{T}      # demand from surface water [mm Δt⁻¹]
+    groundwater_demand::Vector{T}       # demand from groundwater [mm Δt⁻¹]
+end
+
+"Initialize water demand variables"
+function DemandVariables(T::Type{<:AbstractFloat}, n::Int)
+    return DemandVariables{T}(;
+        irri_demand_gross = zeros(T, n),
+        nonirri_demand_gross = zeros(T, n),
+        total_gross_demand = zeros(T, n),
+        surfacewater_demand = zeros(T, n),
+        groundwater_demand = zeros(T, n),
+    )
+end
```

Comment:
I still think things like this can be squashed into one by using default field values allowed by `@with_kw`, or is there a specific reason you're not doing that?

Reply:
Yes, not a specific reason, I would say it is a matter of taste.